### ISSUE TITLE
how can deal with the issue about input with varying dimensions, not using resize or padding zeros?

### ISSUE THREAD

**Issue by jasonTuZq** [fchollet/keras#4539]
Hi, how can deal with the issue about input with varying dimensions, like images with different sizes, but not using resize or padding zeros? I meant to train network with cnn and rnn, like the following:
# conv block 1
model.add(Convolution1D(64,5, input_shape=X_train.shape[1:]))
model.add(Activation('relu'))
model.add(MaxPooling1D(pool_length=3))
model.add(Dropout(0.25))

# conv block 2
model.add(Convolution1D(32,5))
model.add(Activation('relu'))
model.add(MaxPooling1D(pool_length=3))
model.add(Dropout(0.25))

# GRU block 1 & 2
model.add(GRU(16,return_sequences=True))
model.add(GRU(16,return_sequences=False))

Since both cnn and rnn can handle inputs with varying dimension, i thought i could feed inputs with varying dimension into this network, but not knowing how when we have to specify the input_shape for the first layer. 
Can anyone help me?
Thanks!

Please make sure that the boxes below are checked before you submit your issue. Thank you!

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by phdowling** [comment]
Yes, both layers can handle varying sequence lengths, but you will still need to specify one fixed size for the model due to Keras internal implementation. For RNNs, you could alternatively specify a sequence length of 1 and then use the `stateful` flag to manage sequence resets, etc. However, you can also use masking to "hide" 0 inputs from the RNN, and this is what I would recommend you do.

For CNN, masking will not work currently AFAIK. I think there are some problems with making this work in the general case, however your model performance should probably not suffer too much if you just pad your data, the conv filters should probably adjust to this during training.
